### DISCUSSION TITLE
Configuration Of QR Parameters

### DISCUSSION THREAD

**Discussion by sukith** [chillerlan/php-qrcode#190] [Q&A]
Kindly would like to request some answers to the following.

1. Does it make sense to use an absolute pixel size when combining front and background colors or transparency with a logo with Logospace around it? or does it create better quality to "scale" a little bit OVER 2046 Pixels and then use it in this size?

2. Is there a possibility to use Errorlevel Q with a logo with logo space?

3. is it possible to generate an output image with dynamic width & height? We have tried to this extended function called “LogoOptions” and unfortunately, it's not working. Hence, Could you please advise us if is there any solution you had to this issue?

My Index File:
________________________________________________________

require_once('./../vendor/autoload.php');
include 'separateData.php';

$data = $prefix_url;

$options = new LogoOptions(
[
'eccLevel' => QRCode::ECC_H,
'imageBase64' => true,
'logoSpaceHeight' => 12,
'logoSpaceWidth' => 12,
// 'scale' => 10,
'qrCodeHeight' => 500,
'qrCodeWidth' => 500,
'version' => 7,
'outputType' => QRCode::OUTPUT_IMAGE_PNG,
'moduleValues' => $qrCodeOptions,
]
);

$qrOutputInterface = new QRImageWithLogo(
$options,
(new QRCode($options))->getMatrix($data)
);

$qrCode = $qrOutputInterface->dump(
null,
__DIR__.'/../public/img/'.$logo.'.png'
);

$qrCode = str_replace('data:image/svg;base64,', '', $qrCode);
$qrCode = str_replace(' ', '+', $qrCode);
$qrCode = base64_decode($qrCode);
$qrCode = imagecreatefromstring($qrCode);
$qrCode = imagepng($qrCode, __DIR__.'/../public/cache/'.$logo.'.png');

Extended LogoOptions file:
--------------------------------------------------------------------------------------
declare(strict_types=1);

namespace App\QR\Options;

use chillerlan\QRCode\QROptions;

class LogoOptions extends QROptions
{
// size in QR modules, multiply with QROptions::$scale for pixel size
protected int $logoSpaceWidth;
protected int $logoSpaceHeight;

// additional options for logo
protected int $qrCodeHeight;
protected int $qrCodeWidth;

public function __construct(array $options = [])
{
parent::__construct($options);

$this->logoWidth = $options['logoWidth'] ?? 0;
$this->logoHeight = $options['logoHeight'] ?? 0;
$this->qrCodeWidth = $options['qrCodeWidth'] ?? 0;
$this->qrCodeHeight = $options['qrCodeHeight'] ?? 0;
}

public function getLogoWidth(): int
{
return $this->logoWidth;
}

public function getLogoHeight(): int
{
return $this->logoHeight;
}

public function getQrCodeWidth(): int
{
return $this->qrCodeWidth;
}

public function getQrCodeHeight(): int
{
return $this->qrCodeHeight;
}
}

**Comment by codemasher** [comment]
Hey,

> Does it make sense to use an absolute pixel size when combining front and background colors or transparency with a logo with Logospace around it? or does it create better quality to "scale" a little bit OVER 2046 Pixels and then use it in this size?

This is completely up to you. the scale just defines the size of one QR Code pixel (module as per spec). you can use the scale like you suggested to get near the desired image size. Alternatively, you could resize the final image, in which case you may set scale that the generated QR Code is larger, e.g. 1.5x the size of the final resized output.


> Is there a possibility to use Errorlevel Q with a logo with logo space?

No, it is not, because the logo space uses up a lot of the error correction bits and the QR Code may become unreadable. However, I've suggested in another thread to add an option to manually adjust this threshold.


> is it possible to generate an output image with dynamic width & height?

[`imagescale()`](https://www.php.net/manual/function.imagescale.php) is your friend.


Also, i see that you're manually decoding the base64 output to save it to a file. this is not necessary. You can just set `QROptions::$imageBase64` to `false`. Further, `QRCode::render()` has a second parameter that allows you to specifiy a file path where the output should be saved.

**Reply by codemasher** [reply]
This is about how your above code would look like, assuming you have modified the `QRImageWithLogo` class accordingly. The [options container](https://github.com/chillerlan/php-settings-container#readme) uses magic getters and setters, there's no need to manually set these unless you want to check/modify the in- or output values.

```php
class LogoOptions extends QROptions{

	// size in QR modules, multiply with QROptions::$scale for pixel size
	protected int $logoSpaceWidth = 0;
	protected int $logoSpaceHeight = 0;

	// additional options for logo
	protected int $qrCodeHeight = 0;
	protected int $qrCodeWidth = 0;

	// path to logo
	protected string $logo;

	// check logo
	protected function set_logo(string $logo):void{

		if(!file_exists($logo) || !is_readable($logo)){
			throw new QRCodeException('invalid logo');
		}

		// @todo: validate image type

		$this->logo = $logo;
	}

}


$options = new LogoOptions([
	'eccLevel'        => EccLevel::H,
//	'imageBase64'     => false, // not necessary for file output
	'logoSpaceHeight' => 12,
	'logoSpaceWidth'  => 12,
	// 'scale' => 10,
	'qrCodeHeight'    => 500,
	'qrCodeWidth'     => 500,
	'version'         => 7,
	'outputType'      => QROutputInterface::CUSTOM,
	'outputInterface' => QRImageWithLogo::class,
	'moduleValues'    => $qrCodeOptions,
]);


(new QRCode($options))->render($data, __DIR__.'/../public/img/'.$logo.'.png');
```

**Reply by sukith** [reply]
Hi Codesmasher, 

We followed your instruction, and what we found is the transparency works with all other attributes, but when we try to resize the png it adds a white background to the image. Do you know why this is happening or do you have a better approach towards resizing, like utilizing another library?

Best regards,
Sukith

**Reply by codemasher** [reply]
Hi, you might need to re-apply `imagecolortransparent()` after the resize operation. See:
https://github.com/chillerlan/php-qrcode/blob/284c42db884b98b2453e39a7e10a7a04239d0019/src/Output/QRGdImage.php#L169-L170
https://github.com/chillerlan/php-qrcode/blob/284c42db884b98b2453e39a7e10a7a04239d0019/src/Output/QRGdImage.php#L209-L225

**Comment by sukith** [comment]
Hi Codesmasher,

I highly appreciate your prompt replies, we tried this with your normal library downloaded through composer and it says that GDImage script is missing. Do you know any reason for this?

Thanks!
Best regards,
Sukith

**Reply by codemasher** [reply]
Hi, please use the v5.0 beta! https://github.com/chillerlan/php-qrcode/releases/tag/5.0-beta

**Reply by sukith** [reply]
Hi Codemasher,

We used this and we got our QR code to resize as mentioned in the previous thread we used  $this->setTransparencyColor();  after resizing. But there is a problem with our implementation.

1. Our implementation requires the QR code to get resized in a different server after generation.
2. When this happens the QR code generated through chillerlan v5.0 beta again gets a white background after resizing. We don't have access to the other service as its a merchandising platform and we are sending this transparent QR code to be printed on t-shirts.
3. We also tried a different png from a different library(not QR code) and sent it to this third-party service it perfectly works.

Are there any solutions that you see, you can give me how to solve my current problem? The QR code generation is super good and it works flawlessly. It is just whenever we resize the generated QR code elsewhere the transparency goes off.

Best regards,
Sukith

**Reply by codemasher** [reply]
This is a a general issue not within the focus of this library. You will need to check with other resources to solve your resizing issue. Best of luck!
